### ISSUE TITLE
Flatpak: Run from systemd service unit

### ISSUE THREAD

**Issue by salim-b** [projecthamster/hamster#684]
# Goal

I've tried to [hack together](https://gitlab.com/salim_b/bash/hamster_tracking) an alternative for the [removed GNOME session idle detection](https://github.com/projecthamster/hamster/issues/638) running the [Hamster Flatpak app](https://github.com/projecthamster/hamster#flatpak) (GH actions [artifact](https://github.com/projecthamster/hamster/suites/3007492048/artifacts/68043068)).

The approach basically consists of

1. a bash script that regularly polls user session inactivity from GNOME's window manager Mutter and stops time tracking if the specified threshold is reached and resumes time tracking as soon as user activity is detected again. This works fine. The script is found [here](https://gitlab.com/salim_b/bash/hamster_tracking/-/blob/master/hamster_suspend_on_idle.sh).

2. some more scripts to run from systemd service units with the goal of a) stopping time tracking when the system is powered off and b) suspending and resuming time tracking before and after the system is put to sleep. This doesn't work as supposed yet.

# Details

For the second point, consider the following bash script `/usr/local/bin/hamster_tracking/hamster_stop.sh` and a systemd service unit `/etc/systemd/system/hamster_stop.service` that executes the former before the system powers off:

<details>
<summary><b><code>hamster_stop.sh</code></b></summary>

```h
#!/bin/bash

# abort if Hamster Flatpak is not available
if ! command -v flatpak >/dev/null || ! flatpak list --app | grep -q 'org\.gnome\.Hamster'; then
  echo "The Hamster Flatpak is required in order for this script to work, but is not installed. Aborting."
  exit 1
fi

# stop current Hamster activity
current_activity=`flatpak run org.gnome.Hamster current`

if [[ ${current_activity} != "No activity" ]] ; then
  flatpak run org.gnome.Hamster stop
fi
```
</details

---

<details>
<summary><b><code>hamster_stop.service</code></b></summary>

```ini
[Unit]
Description=Stop Hamster activity on system power off
DefaultDependencies=no
Conflicts=reboot.target
Before=shutdown.target

[Service]
Type=oneshot
User=salim
Environment=DISPLAY=:0
ExecStart=/usr/local/bin/hamster_tracking/hamster_stop.sh

[Install]
WantedBy=shutdown.target
```
</details

---

# Test

The above service unit can be tested using:

```sh
# reload and reenable the unit in case of recent edits
sudo systemctl daemon-reload
sudo systemctl reenable hamster_stop.service

# run the unit
sudo systemctl start hamster_stop.service
```

# Problem

My main problem now is: Although running `flatpak run org.gnome.Hamster current` inside a system service unit with `User=salim` does appear to work as long as `Environment=DISPLAY=:0` is set, **it doesn't return the proper activity value**. It always returns nothing while it should return a line like `YYYY-MM-DD HH:MM ACTIVITY@CATEGORY HH:MM` or `No activity`.

Thus my main question (maybe @tchernobog can help?): **How can I properly run the Hamster Flatpak app from within a systemd system service unit?**

Thanks in advance for any help!

**Comment by salim-b** [comment]
Meanwhile I could solve the original problem and thus completely overhauled my initial post above as well as the issue title.

**Comment by tchernobog** [comment]
Assuming that what you want is to run hamster for your user "salim", you should be at least using user units in `$HOME/.config/systemd/user`, and call `systemctl` with the `--user` flag. That would remove "sudo" from the equation. And also take in account the use case of the user logging out without rebooting.

I also think a better approach would be to use a DBus message via `dbus-send`. I believe that would simplify the logic of what you are trying to do.

**Comment by salim-b** [comment]
> Assuming that what you want is to run hamster for your user "salim", you should be at least using user units in $HOME/.config/systemd/user, and call systemctl with the --user flag. That would remove "sudo" from the equation.

Yeah, initially I thought the same. But this is not possible in the current systemd world – at least not for sleep/shutdown/etc.-related stuff – because ["'user' instances currently have no way to depend on systemwide services"](https://unix.stackexchange.com/questions/149959/how-to-run-systemd-user-service-to-trigger-on-sleep-aka-suspend-hibernate/174837#174837).

Besides, `sudo` is not the issue here. :)

> I also think a better approach would be to use a DBus message via dbus-send. I believe that would simplify the logic of what you are trying to do. I did not try, but I believe org.gnome.Hamster.StopTracking(Int32 end_time) should do what you need.

Awesome, thanks for pointing out this alternative! I'll test this.

Is there also a D-Bus message/call to query the currently running Hamster activity? Or – more generally – where is Hamster's D-Bus service API documented?

**Comment by salim-b** [comment]
> And also take in account the use case of the user logging out without rebooting.

Good point. I didn't think about this yet (rather irrelevant for my use case since I'm the only user on my system).

**Comment by GeraldJansen** [comment]
See https://github.com/projecthamster/hamster/blob/master/src/hamster-service.py for Hamster's D-Bus service API.
You can check the last element of  GetTodaysFactsJSON(): no end time means  the last activity is ongoing. Anyway StopTracking will do nothing if there is no ongoing activity, so you can just call it without checking (use end_time 0 for current time).

**Comment by GeraldJansen** [comment]
Perhaps you could use dbus-monitor to listen for org.freedesktop.login1 PrepareForShutdown and PrepareForSleep signals? Maybe they get sent before your flatpak hamster process gets killed. (Warning: I know nothing about this stuff!)

**Comment by tchernobog** [comment]
I don't think this ticket is a proper issue with Hamster itself. I'd like to see it closed.
